internal/overlord/servstate/handlers.go

```diff
@@ -51,183 +54,528 @@ var (
+// and it sets s.cmd and other relevant fields.
+func (s *service) startInternal() error {
+	args, err := shlex.Split(s.config.Command)
 	if err != nil {
 		// Shouldn't happen as it should have failed on parsing, but
 		// it does not hurt to double check and report.
 		return fmt.Errorf("cannot parse service command: %s", err)
 	}
-	cmd := exec.Command(args[0], args[1:]...)
-	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
+	s.cmd = exec.Command(args[0], args[1:]...)
+	s.cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
 
-	// Start as another user if specified in plan
-	uid, gid, err := osutil.NormalizeUidGid(service.UserID, service.GroupID, service.User, service.Group)
+	// Start as another user if specified in plan.
+	uid, gid, err := osutil.NormalizeUidGid(s.config.UserID, s.config.GroupID, s.config.User, s.config.Group)
 	if err != nil {
 		return err
 	}
 	if uid != nil && gid != nil {
-		setCmdCredential(cmd, &syscall.Credential{
+		setCmdCredential(s.cmd, &syscall.Credential{
 			Uid: uint32(*uid),
 			Gid: uint32(*gid),
 		})
 	}
 
-	// Pass service description's environment variables to child process
-	cmd.Env = os.Environ()
-	for k, v := range service.Environment {
-		cmd.Env = append(cmd.Env, k+"="+v)
+	// Pass service description's environment variables to child process.
+	s.cmd.Env = os.Environ()
+	for k, v := range s.config.Environment {
+		s.cmd.Env = append(s.cmd.Env, k+"="+v)
 	}
-	logBuffer := servicelog.NewRingBuffer(maxLogBytes)
+
+	// Set up stdout and stderr to write to log ring buffer.
 	var outputIterator servicelog.Iterator
-	if m.serviceOutput != nil {
-		outputIterator = logBuffer.TailIterator()
+	if s.manager.serviceOutput != nil {
+		outputIterator = s.logs.HeadIterator(0)
 	}
+	logWriter := servicelog.NewFormatWriter(s.logs, s.config.Name)
+	s.cmd.Stdout = logWriter
+	s.cmd.Stderr = logWriter
 
-	logWriter := servicelog.NewFormatWriter(logBuffer, req.Name)
-	cmd.Stdout = logWriter
-	cmd.Stderr = logWriter
-	err = cmd.Start()
+	// Start the process!
+	err = s.cmd.Start()
 	if err != nil {
 		if outputIterator != nil {
 			_ = outputIterator.Close()
 		}
-		_ = logBuffer.Close()
+		_ = s.logs.Close()
 		return fmt.Errorf("cannot start service: %v", err)
 	}
+	startTime, _ := s.config.ParseStartTime() // ignore error; it's already been validated
+	s.resetTimer = time.AfterFunc(startTime, func() { logError(s.startTimeElapsed()) })
 
-	active := &activeService{
-		originalPlan: service.Copy(),
-		cmd:          cmd,
-		done:         make(chan struct{}),
-		logBuffer:    logBuffer,
-	}
-	m.services[req.Name] = active
+	// Start a goroutine to wait for the process to finish.
+	done := make(chan struct{})
 	go func() {
-		active.err = cmd.Wait()
-		_ = active.logBuffer.Close()
-		close(active.done)
+		waitErr := s.cmd.Wait()
+		close(done)
+		err := s.exited(waitErr)
+		if err != nil {
+			logger.Noticef("Cannot execute exited action: %v", err)
```

Comment:
This message needs to be a bit more user friendly so people at the terminal can tell what it means.

Reply:
Note that this is not actually something a user should ever see, as the errors `exited()` returns are both internal "shouldn't happen" error. I've tweaked the wording slightly, though. So if we did have a bug they'd see something like:

```
Cannot transition state after service exited: internal error: exited invalid in state "backoff"
```